File: python/pygraph/nativepy/astar.py
# astar_path.pyx

from heapq import heappop, heappush
from itertools import count
from pygraph.nativepy.multidigraph import MultiDiGraph


class NodeNotFound(Exception):
    """Exception raised if node not present in graph.
    """
    pass


class NoPath(Exception):
    """Exception raised if no path can be found.
    """
    pass


def _weight_function(G: MultiDiGraph , weight):
    if callable(weight):
        return weight
    elif weight is None:
        return lambda u, v, d: 1.0
    elif type(weight) == str:
        def weight_func(g, u, v, key, key_previous, v_previous):
            d = g[u][v][key]
            return d.get(weight, 1.0)
        return weight_func
    else:
        raise TypeError("weight must be None, a string or a function")


def base_heuristic(u, u_data, v, v_data):
    return 0


def astar_path(G: MultiDiGraph, source, target, heuristic=None, weight="weight") -> list:
    if source not in G or target not in G:
        msg = f"Either source {source} or target {target} is not in G"
        raise NodeNotFound(msg)

    weight = _weight_function(G, weight)
    if heuristic is None:
        heuristic = base_heuristic
    # The queue stores priority, node, cost to reach, parent and key.
    # Uses Python heapq to keep in priority order.
    # Add a counter to the queue to prevent the underlying heap from
    # attempting to compare the nodes themselves. The hash breaks ties in the
    # priority and is guaranteed unique for all nodes in the graph.
    c = count()
    queue = [(0, next(c), source, 0, None, None)]
    # Maps enqueued nodes to distance of discovered paths and the
    # computed heuristics to target. We avoid computing the heuristics
    # more than once and inserting the node into the queue too many times.
    enqueued = {}
    # Maps explored nodes to parent closest to the source.
    explored = {}
    while queue:
        # Pop the smallest item from queue.
        _, __, curnode, dist, parent, key = heappop(queue)
        if curnode == target:
            path = [
                (parent, curnode, key)
            ]
            node = parent
            while node is not None:
                parent = node
                node, key = explored[node]
                if node is not None:
                    path.append(
                        (node, parent, key)
                    )
            path.reverse()

            return path

        if curnode in explored:
            # Do not override the parent of starting node
            if explored[curnode] == (None, None):
                continue

            # Skip bad paths that were enqueued before finding a better one
            qcost, h = enqueued[curnode]
            if qcost < dist:
                continue

        explored[curnode] = (parent, key)
        key_previous = key
        for neighbor, w in G[curnode].items():
            # loop over all existing edges
            for key in w.keys():
                ncost = dist + weight(
                    g=G,
                    u=curnode,
                    v=neighbor,
                    key=key,
                    key_previous=key_previous,
                    v_previous=parent,
                )
                if neighbor in enqueued:
                    qcost, h = enqueued[neighbor]
                    # if qcost <= ncost, a less costly path from the
                    # neighbor to the source was already determined.
                    # Therefore, we won't attempt to push this neighbor
                    # to the queue
                    if qcost <= ncost:
                        continue
                else:
                    h = heuristic(
                        neighbor,
                        G.get_node_data(neighbor),
                        target,
                        G.get_node_data(target),
                    )
                enqueued[neighbor] = ncost, h
                heappush(queue, (ncost + h, next(c), neighbor, ncost, curnode, key))

    raise NoPath("No path found between %s and %s." % (source, target))

File: cpp/src/astar.cpp
#include <unordered_map>
#include <string>
#include <utility>
#include <vector>
#include <queue>
#include <stdexcept>
#include <iostream>
#include <static/astar.h>
#include <static/multidigraph.h>

// Exception Classes Implementation
NodeNotFound::NodeNotFound(const std::string& message) : std::runtime_error(message) {}

NoPath::NoPath(const std::string& message) : std::runtime_error(message) {}

// DummyHeuristic Implementation
double DummyHeuristic::call(const std::string& u, const NodeData& u_data,
                            const std::string& v, const NodeData& v_data) {
    return 0.0;
}

// DummyWeightFunction Implementation
double DummyWeightFunction::call(MultiDiGraph& g, const std::string& u, const std::string& v,
                                 const std::string& key, const std::string& key_previous, const std::string& v_previous) {
    return 1.0;
}

// GetWeightFunction Implementation
GetWeightFunction::GetWeightFunction(const std::string& weight) : weight(weight) {}

double GetWeightFunction::call(MultiDiGraph& g, const std::string& u, const std::string& v,
                               const std::string& key, const std::string& key_previous, const std::string& v_previous) {
    EdgeData edge_data = g.adj(u)[v][key];
    return edge_data[weight];
}

// QueueItem Implementation
QueueItem::QueueItem(double _priority, int _counter, std::string  _node, double _dist,
                     std::string  _parent, std::string  _key)
        : priority(_priority), counter(_counter), node(std::move(_node)), dist(_dist),
          parent(std::move(_parent)), key(std::move(_key)) {}

bool QueueItem::operator<(const QueueItem& other) const {
    return priority > other.priority;
}


std::vector<std::tuple<std::string, std::string, std::string>>
astar_path(MultiDiGraph& G, const std::string& source, const std::string& target,
           HeuristicFunction* heuristic, WeightFunction* weight) {

    std::unordered_map<std::string, std::pair<double, double>> enqueued;
    std::unordered_map<std::string, std::pair<std::string, std::string>> explored;
    std::pair<std::string, std::string> empty_node("", "");

    if (!G.has_node(source) || !G.has_node(target)) {
        throw NodeNotFound("Either source " + source + " or target " + target + " is not in G");
    }

    WeightFunction* weight_func = weight ? weight : new DummyWeightFunction();
    HeuristicFunction* heuristic_func = heuristic ? heuristic : new DummyHeuristic();

    std::priority_queue<QueueItem> queue;
    int counter = 0;
    queue.push(QueueItem(0.0, counter, source, 0.0, std::string(""), std::string("")));

    while (!queue.empty()) {
        std::cout << queue.size() << std::endl;
        const QueueItem& topItem = queue.top();
        queue.pop();
        std::string curnode = topItem.node;
        double dist = topItem.dist;
        std::string parent = topItem.parent;
        std::string key = topItem.key;
        if (curnode == target) {
            std::vector<std::tuple<std::string, std::string, std::string>> path;
            path.emplace_back(parent, curnode, key);
            std::string node = parent;
            while (!node.empty()) {
                std::cout << "Node: " << node << std::endl;
                parent = node;
                node = explored[node].first;
                key = explored[node].second;
                path.emplace_back(node, parent, key);
            }
            std::reverse(path.begin(), path.end());
            return path;
        }

        auto [it, inserted] = explored.insert_or_assign(curnode, std::make_pair(parent, key));
        if (!inserted && it->second != empty_node && enqueued[curnode].first < dist) {
            continue;
        }

        explored[curnode] = {parent, key};
        std::string key_previous = key;

        for (const auto& [neighbor, w] : G.adj(curnode)) {
            for (const auto& [edge_key, data] : w) {
                double ncost = dist + weight_func->call(G, curnode, neighbor, edge_key, key_previous, parent);
                auto enqueuedIter = enqueued.find(neighbor);
                if (enqueuedIter != enqueued.end() && enqueuedIter->second.first <= ncost) {
                    continue;
                } else {
                    const NodeData& neighbor_node_data = G.get_node_data(neighbor);
                    const NodeData& target_node_data = G.get_node_data(target);
                    double h = heuristic_func->call(neighbor, neighbor_node_data, target, target_node_data);
                    enqueued[neighbor] = {ncost, h};
                    counter++;
                    queue.push(QueueItem(ncost + h, counter, neighbor, ncost, curnode, edge_key));
                }
            }
        }
    }

    throw NoPath("No path found between " + source + " and " + target);
}
